File: cs/days/day19.cs
namespace Shunty.AoC.Days;

// https://adventofcode.com/2024/day/19 - Linen Layout

public class Day19 : AocDaySolver
{
    public int DayNumber => 19;
    public string Title => "Linen Layout";

    public async Task Solve()
    {
        var input = (await File.ReadAllLinesAsync(AocUtils.FindInputFile(DayNumber))).ToList();
        //var input = (await Task.FromResult(TestInput.Trim())).Split('\n').ToList();
        var towels = input[0].Trim().Split(", ").AsReadOnly();
        var patterns = input[2 ..].Select(s => s.Trim());

        var part1 = patterns.Sum(p => CanMakePattern(towels, p));
        this.ShowDayResult(1, part1);

        Dictionary<string, long> cache = [];
        long part2 = patterns.Sum(p => PatternCombinations(towels, p, cache));
        this.ShowDayResult(2, part2);
    }

    /// <summary>
    /// This is the original fn for Part 1. Turns out not to be suitable for part 2 (surprise)
    /// and part 2 was so much more straightforward with a bit of recursion and a cache anyway.
    /// I could substitute the part 2 solution for this but why bother now.
    private int CanMakePattern(IReadOnlyCollection<string> towels, string pattern)
    {
        Queue<int> q = [];
        HashSet<int> matchedlengths = [];
        var ps = pattern.AsSpan();
        foreach (var towel in towels)
        {
            if (ps.StartsWith(towel))
                q.Enqueue(towel.Length);
        }
        var maxlen = towels.Max(t => t.Length);
        var plen = ps.Length;
        while (q.Count > 0)
        {
            var len = q.Dequeue();
            foreach (var towel in towels)
            {
                var newlen = len + towel.Length;
                if (newlen > plen || matchedlengths.Contains(newlen))
                    continue;
                if (ps[len..].StartsWith(towel))
                {
                    if (newlen == plen)
                        return 1;
                    q.Enqueue(newlen);
                    matchedlengths.Add(newlen);
                }
            }
        }
        return 0;
    }

    private long PatternCombinations(IReadOnlyCollection<string> towels, string pattern, Dictionary<string, long> cache)
    {
        if (cache.TryGetValue(pattern, out var result))
            return result;

        if (string.IsNullOrWhiteSpace(pattern))
            return 1;

        result = 0;
        foreach (var towel in towels)
        {
            if (pattern.StartsWith(towel))
            {
                result += PatternCombinations(towels, pattern[towel.Length..], cache);
            }
        }
        cache[pattern] = result;
        return result;
    }

    private const string TestInput = """
    r, wr, b, g, bwu, rb, gb, br

    brwrr
    bggr
    gbbr
    rrbgbr
    ubwu
    bwurrg
    brgr
    bbrgwb
    """;
}

File: go/days/day19.go
package days

import (
	aoc "aoc2024/aocutils"
	"strings"
)

type day19 struct {
	day int
}

func Day19() {
	d := &day19{19}
	d.Run()
}

func (d *day19) Run() {
	input, err := aoc.GetDayInputLines(d.day)
	aoc.CheckError(err)

	towels := strings.Split(input[0], ", ")
	patterns := input[2:]

	aoc.DayHeader(d.day)
	part1, part2 := d.findDesigns(towels, patterns)
	aoc.PrintResult(1, part1)
	aoc.PrintResult(2, part2)
}

// Combine part1 and part2. P1 only requires the number of patterns that are
// achievable and P2 needs to know how many ways we can make each one.
func (d *day19) findDesigns(towels []string, patterns []string) (int, int64) {
	cache := make(map[string]int64)
	count := 0
	var perms int64 = 0
	for _, pattern := range patterns {
		dp := d.findDesignPerms(towels, pattern, cache)
		if dp > 0 {
			count++
		}
		perms += dp
	}
	return count, perms
}

// This cries out for recursion and memoization (a cache)
func (d *day19) findDesignPerms(towels []string, pattern string, cache map[string]int64) int64 {
	if pattern == "" {
		return 1
	}
	if v, ok := cache[pattern]; ok {
		return v
	}

	var result int64 = 0
	for _, towel := range towels {
		if strings.HasPrefix(pattern, towel) {
			result += d.findDesignPerms(towels, pattern[len(towel):], cache)
		}
	}
	cache[pattern] = result
	return result
}
